### DISCUSSION TITLE
util:User unit tests missing 'Domain' and no length / validity checks on entries

### DISCUSSION THREAD

**Discussion by bevanweiss** [wixtoolset/issues#8570] [WiX Development]
Whilst I've been looking to add the Create / Remove User Group capability to WiX I've been going through and largely copying the Wix4User code (since it's largely similar to how I think User Groups should work).

During this I've notice a few things:

1. There was a missing localization for a string in the error codes potentially returned.  I've raised a Pull Request (https://github.com/wixtoolset/wix/pull/529) to fix that.
2. There's no check on the length of data that can be supplied to the various fields.  e.g. the Wix4User Name column is a string of 255 characters, but in the Win32API only 20 characters are supported.  There's also no check for invalid characters during compilation (that I can see).  Should this be changed?  I could easily add in such constraints to a Pull Request (separate from my User Group Pull Request) if that would be desired.  I'm unsure how this would tie in to runtime Properties however... obviously they'd need to exempted from compile time errors...
> User account names are limited to 20 characters and group names are limited to 256 characters. In addition, account names cannot be terminated by a period and they cannot include commas or any of the following printable characters: ", /, , [, ], :, |, <, >, +, =, ;, ?, *. Names also cannot include characters in the range 1-31, which are nonprintable.
3. There doesn't appear to be any 'domain' entries in the current Unit Test for the Custom Action data transfer.  I could wrap this into a Pull Request alongside the constraints for the User fields if desired.
``` csharp
        public void CanCreateUserAccountWithComment()
        {
            var folder = TestData.Get(@"TestData\CreateUser");
            var build = new Builder(folder, typeof(UtilExtensionFactory), new[] { folder });

            var results = build.BuildAndQuery(Build, "Wix4User");
            WixAssert.CompareLineByLine(new[]
            {
                "Wix4User:TEST_USER00\tComponent1\ttestName00\t{DOMAIN SHOULD BE HERE}\ttest123!@#\tTest Comment 1\t1792",
                "Wix4User:TEST_USER01\tComponent1\ttestName01\t\ttest123!@#\tTest Comment 1\t1796",
```

**Comment by robmen** [comment]
1. We're not generally a fan of using generated localization. However, something needs to be in place to extract the localization string so there probably isn't a better option (which is probably why it didn't happen before).
2. Right now, the expectation is that the Win32 APIs will fail with a reasonable error code, so we don't try to duplicate the checks. Duplicating Windows validation is fraught with peril and should only be done when the value is high.
3. Improving tests is pretty much always a good thing.
